File: src/enrich.rs
use std::collections::HashSet;

use serde_json::{Map, Value};
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

use crate::error::AppError;

pub struct EnrichConfig {
    pub source: String,
    pub result: String,
}

struct EntryInfo {
    text: Option<String>,
    start_ms: Option<i64>,
    end_ms: Option<i64>,
    start_formatted: Option<String>,
    end_formatted: Option<String>,
    id: Option<String>,
}

/// CSV column indices: id,text,start_ms,end_ms,start_formatted,end_formatted
const COL_ID: usize = 0;
const COL_TEXT: usize = 1;
const COL_START_MS: usize = 2;
const COL_END_MS: usize = 3;
const COL_START_FMT: usize = 4;
const COL_END_FMT: usize = 5;

fn non_empty(s: &str) -> Option<String> {
    if s.is_empty() { None } else { Some(s.to_string()) }
}

/// Build lookup from a preprocessed CSV file (canonical format).
fn build_entry_lookup(source: &str) -> crate::error::Result<Vec<EntryInfo>> {
    let file = File::open(Path::new(source)).map_err(|e| AppError::FileOpen {
        path: source.to_string(),
        source: e,
    })?;
    let mut rdr = csv::ReaderBuilder::new()
        .has_headers(true)
        .from_reader(BufReader::new(file));
    let mut entries = Vec::new();

    for (line_num, result) in rdr.records().enumerate() {
        let record = result.map_err(|e| AppError::Generic(format!("line {}: {e}", line_num + 1)))?;

        entries.push(EntryInfo {
            id: record.get(COL_ID).and_then(non_empty),
            text: record.get(COL_TEXT).and_then(non_empty),
            start_ms: record.get(COL_START_MS).and_then(|s| s.parse().ok()),
            end_ms: record.get(COL_END_MS).and_then(|s| s.parse().ok()),
            start_formatted: record.get(COL_START_FMT).and_then(non_empty),
            end_formatted: record.get(COL_END_FMT).and_then(non_empty),
        });
    }

    Ok(entries)
}

fn inject_entry_info(ts: &mut Map<String, Value>, info: &EntryInfo) {
    if let Some(s) = info.start_ms {
        ts.insert("start_ms".to_string(), Value::from(s));
    }
    if let Some(e) = info.end_ms {
        ts.insert("end_ms".to_string(), Value::from(e));
    }
    if let Some(sf) = &info.start_formatted {
        ts.insert("start_formatted".to_string(), Value::from(sf.clone()));
    }
    if let Some(ef) = &info.end_formatted {
        ts.insert("end_formatted".to_string(), Value::from(ef.clone()));
    }
    if let Some(id) = &info.id {
        ts.insert("id".to_string(), Value::from(id.clone()));
    }
    if let Some(text) = &info.text {
        ts.insert("text".to_string(), Value::from(text.clone()));
    }
}

fn enrich_value(value: &mut Value, lookup: &[EntryInfo]) {
    match value {
        Value::Object(map) => {
            // If this object has a start_index, inject timestamp and id fields
            if let Some(idx) = map.get("start_index").and_then(|v| v.as_u64())
                && let Some(info) = lookup.get(idx as usize)
            {
                inject_entry_info(map, info);
            }

            // For arrays whose elements start with [index, ...], inject timestamps.
            // Covers: "indices" (exact_duplicates), "members" (near_duplicates),
            // "entry_indices" (ngrams).
            for key in ["indices", "members", "entry_indices"] {
                if let Some(Value::Array(arr)) = map.get(key) {
                    let ts_array: Vec<Value> = arr
                        .iter()
                        .filter_map(|v| {
                            let a = v.as_array()?;
                            let idx = a.first()?.as_u64()?;
                            Some(idx)
                        })
                        .map(|idx| {
                            let mut ts = Map::new();
                            ts.insert("index".to_string(), Value::from(idx));
                            if let Some(info) = lookup.get(idx as usize) {
                                inject_entry_info(&mut ts, info);
                            }
                            Value::Object(ts)
                        })
                        .collect();
                    map.insert("index_timestamps".to_string(), Value::Array(ts_array));
                }
            }

            // Recurse into all values
            for v in map.values_mut() {
                enrich_value(v, lookup);
            }
        }
        Value::Array(arr) => {
            for v in arr {
                enrich_value(v, lookup);
            }
        }
        _ => {}
    }
}

/// Collect all entry indices that are covered by any repetition pattern.
fn collect_repeated_indices(result_json: &Value) -> HashSet<usize> {
    let mut repeated = HashSet::new();

    // From exact_duplicates: indices are [[index, id], ...]
    if let Some(Value::Array(groups)) = result_json.get("exact_duplicates") {
        for group in groups {
            if let Some(Value::Array(indices)) = group.get("indices") {
                for entry in indices {
                    if let Some(idx) = entry.as_array().and_then(|a| a.first()?.as_u64()) {
                        repeated.insert(idx as usize);
                    }
                }
            }
        }
    }

    // From near_duplicates: members are [[index, id, text], ...]
    if let Some(Value::Array(clusters)) = result_json.get("near_duplicates") {
        for cluster in clusters {
            if let Some(Value::Array(members)) = cluster.get("members") {
                for member in members {
                    if let Some(idx) = member.as_array().and_then(|a| a.first()?.as_u64()) {
                        repeated.insert(idx as usize);
                    }
                }
            }
        }
    }

    // From repeated_sequences: each occurrence covers [start_index, start_index + length)
    if let Some(Value::Array(seqs)) = result_json.get("repeated_sequences") {
        for seq in seqs {
            let length = seq.get("length").and_then(|v| v.as_u64()).unwrap_or(0) as usize;
            if let Some(Value::Array(occurrences)) = seq.get("occurrences") {
                for occ in occurrences {
                    if let Some(start) = occ.get("start_index").and_then(|v| v.as_u64()) {
                        for offset in 0..length {
                            repeated.insert(start as usize + offset);
                        }
                    }
                }
            }
        }
    }

    // From near_duplicate_sequences: same range logic
    if let Some(Value::Array(seqs)) = result_json.get("near_duplicate_sequences") {
        for seq in seqs {
            let length = seq.get("length").and_then(|v| v.as_u64()).unwrap_or(0) as usize;
            if let Some(Value::Array(occurrences)) = seq.get("occurrences") {
                for occ in occurrences {
                    if let Some(start) = occ.get("start_index").and_then(|v| v.as_u64()) {
                        for offset in 0..length {
                            repeated.insert(start as usize + offset);
                        }
                    }
                }
            }
        }
    }

    repeated
}

fn build_segment(lookup: &[EntryInfo], start: usize, end: usize, is_repeated: bool) -> Value {
    let mut seg = Map::new();
    seg.insert(
        "type".to_string(),
        Value::from(if is_repeated { "repeated" } else { "unique" }),
    );
    seg.insert("start_index".to_string(), Value::from(start));
    seg.insert("end_index".to_string(), Value::from(end));
    seg.insert("entry_count".to_string(), Value::from(end - start + 1));

    let texts: Vec<Value> = (start..=end)
        .filter_map(|i| lookup.get(i)?.text.as_deref())
        .map(Value::from)
        .collect();
    seg.insert("texts".to_string(), Value::Array(texts));

    // Timestamps: start from first entry, end from last entry
    let first = lookup.get(start);
    let last = lookup.get(end);
    seg.insert(
        "start_ms".to_string(),
        Value::from(first.and_then(|e| e.start_ms).unwrap_or(0)),
    );
    seg.insert(
        "end_ms".to_string(),
        Value::from(last.and_then(|e| e.end_ms).unwrap_or(0)),
    );
    if let Some(sf) = first.and_then(|e| e.start_formatted.as_ref()) {
        seg.insert("start_formatted".to_string(), Value::from(sf.clone()));
    }
    if let Some(ef) = last.and_then(|e| e.end_formatted.as_ref()) {
        seg.insert("end_formatted".to_string(), Value::from(ef.clone()));
    }

    Value::Object(seg)
}

pub fn run_extract_unique(config: &EnrichConfig) -> crate::error::Result<()> {
    let lookup = build_entry_lookup(&config.source)?;
    let total = lookup.len();

    let result_file = File::open(&config.result).map_err(|e| AppError::FileOpen {
        path: config.result.clone(),
        source: e,
    })?;
    let result_json: Value = serde_json::from_reader(BufReader::new(result_file))?;

    let repeated = collect_repeated_indices(&result_json);

    eprintln!(
        "{} / {} entries covered by repetition patterns",
        repeated.len(),
        total
    );

    // Walk entries and group consecutive same-type indices into segments
    let mut segments: Vec<Value> = Vec::new();
    if total > 0 {
        let mut seg_start = 0usize;
        let mut seg_repeated = repeated.contains(&0);

        for i in 1..total {
            let is_rep = repeated.contains(&i);
            if is_rep != seg_repeated {
                segments.push(build_segment(&lookup, seg_start, i - 1, seg_repeated));
                seg_start = i;
                seg_repeated = is_rep;
            }
        }
        segments.push(build_segment(&lookup, seg_start, total - 1, seg_repeated));
    }

    let json = serde_json::to_string_pretty(&segments)?;
    println!("{json}");
    Ok(())
}

pub fn run_enrich(config: &EnrichConfig) -> crate::error::Result<()> {
    let lookup = build_entry_lookup(&config.source)?;

    eprintln!(
        "Loaded {} entries from source for enrichment lookup",
        lookup.len()
    );

    let result_file = File::open(&config.result).map_err(|e| AppError::FileOpen {
        path: config.result.clone(),
        source: e,
    })?;
    let mut result_json: Value =
        serde_json::from_reader(BufReader::new(result_file))?;

    // Inject total_duration_secs at top level
    if let Value::Object(ref mut map) = result_json
        && let (Some(first), Some(last)) = (lookup.first(), lookup.last())
        && let (Some(s), Some(e)) = (first.start_ms, last.end_ms)
    {
        map.insert(
            "total_duration_secs".to_string(),
            Value::from((e - s) as f64 / 1000.0),
        );
    }

    enrich_value(&mut result_json, &lookup);

    let output = serde_json::to_string_pretty(&result_json)?;
    println!("{output}");
    Ok(())
}

File: repetition_analyzer/enrich.py
"""Timestamp enrichment and unique/repeated segmentation."""

import csv
import json
import sys
from typing import Any


def _non_empty(s: str) -> str | None:
    return s if s else None


def build_entry_lookup(source: str) -> list[dict[str, Any]]:
    """Build lookup from a preprocessed CSV file."""
    entries = []
    with open(source, newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            entries.append({
                "id": _non_empty(row.get("id", "")),
                "text": _non_empty(row.get("text", "")),
                "start_ms": int(row["start_ms"]) if row.get("start_ms") else None,
                "end_ms": int(row["end_ms"]) if row.get("end_ms") else None,
                "start_formatted": _non_empty(row.get("start_formatted", "")),
                "end_formatted": _non_empty(row.get("end_formatted", "")),
            })
    return entries


def _inject_entry_info(target: dict[str, Any], info: dict[str, Any]) -> None:
    for key in ("start_ms", "end_ms", "start_formatted", "end_formatted", "id", "text"):
        if info.get(key) is not None:
            target[key] = info[key]


def enrich_value(value: Any, lookup: list[dict[str, Any]]) -> None:
    """Recursively enrich JSON value with timestamp data from lookup."""
    if isinstance(value, dict):
        # If this object has a start_index, inject timestamp and id fields
        start_index = value.get("start_index")
        if isinstance(start_index, int) and 0 <= start_index < len(lookup):
            _inject_entry_info(value, lookup[start_index])

        # For arrays whose elements start with [index, ...], inject timestamps
        for key in ("indices", "members", "entry_indices"):
            if key in value and isinstance(value[key], list):
                ts_array = []
                for item in value[key]:
                    if isinstance(item, list) and item and isinstance(item[0], int):
                        idx = item[0]
                        ts = {"index": idx}
                        if 0 <= idx < len(lookup):
                            _inject_entry_info(ts, lookup[idx])
                        ts_array.append(ts)
                value["index_timestamps"] = ts_array

        # Recurse into all values
        for v in value.values():
            enrich_value(v, lookup)

    elif isinstance(value, list):
        for v in value:
            enrich_value(v, lookup)


def collect_repeated_indices(result_json: dict[str, Any]) -> set[int]:
    """Collect all entry indices covered by any repetition pattern."""
    repeated: set[int] = set()

    # From exact_duplicates: indices are [[index, id], ...]
    for group in result_json.get("exact_duplicates", []):
        for entry in group.get("indices", []):
            if isinstance(entry, list) and entry:
                repeated.add(entry[0])

    # From near_duplicates: members are [[index, id, text], ...]
    for cluster in result_json.get("near_duplicates", []):
        for member in cluster.get("members", []):
            if isinstance(member, list) and member:
                repeated.add(member[0])

    # From repeated_sequences
    for seq in result_json.get("repeated_sequences", []):
        length = seq.get("length", 0)
        for occ in seq.get("occurrences", []):
            start = occ.get("start_index", 0)
            for offset in range(length):
                repeated.add(start + offset)

    # From near_duplicate_sequences
    for seq in result_json.get("near_duplicate_sequences", []):
        length = seq.get("length", 0)
        for occ in seq.get("occurrences", []):
            start = occ.get("start_index", 0)
            for offset in range(length):
                repeated.add(start + offset)

    return repeated


def _consolidate_repeated(
    repeated: set[int], max_unique_gap: int, min_repeated_island: int,
) -> set[int]:
    """Consolidate repeated indices to reduce fragmentation.

    Pass 1 (close): fill unique gaps of <= max_unique_gap between repeated indices.
    Pass 2 (open): remove contiguous repeated runs of <= min_repeated_island entries.
    """
    if not repeated:
        return repeated

    result = set(repeated)

    # Pass 1: close gaps between nearby repeated indices.
    sorted_rep = sorted(result)
    for a, b in zip(sorted_rep, sorted_rep[1:]):
        gap = b - a - 1
        if 0 < gap <= max_unique_gap:
            result.update(range(a + 1, b))

    # Pass 2: remove small repeated islands.
    sorted_result = sorted(result)
    runs: list[tuple[int, int]] = []
    run_start = sorted_result[0]
    for i in range(1, len(sorted_result)):
        if sorted_result[i] != sorted_result[i - 1] + 1:
            runs.append((run_start, sorted_result[i - 1]))
            run_start = sorted_result[i]
    runs.append((run_start, sorted_result[-1]))

    for start, end in runs:
        if end - start + 1 <= min_repeated_island:
            result.difference_update(range(start, end + 1))

    return result


def _build_segment(lookup: list[dict[str, Any]], start: int, end: int, is_repeated: bool) -> dict[str, Any]:
    texts = [lookup[i]["text"] for i in range(start, end + 1) if lookup[i].get("text")]
    first = lookup[start] if start < len(lookup) else {}
    last = lookup[end] if end < len(lookup) else {}

    seg: dict[str, Any] = {
        "type": "repeated" if is_repeated else "unique",
        "start_index": start,
        "end_index": end,
        "entry_count": end - start + 1,
        "texts": texts,
        "start_ms": first.get("start_ms", 0) or 0,
        "end_ms": last.get("end_ms", 0) or 0,
    }
    if first.get("start_formatted"):
        seg["start_formatted"] = first["start_formatted"]
    if last.get("end_formatted"):
        seg["end_formatted"] = last["end_formatted"]
    return seg


def run_extract_unique(config: dict[str, Any]) -> None:
    lookup = build_entry_lookup(config["source"])
    total = len(lookup)

    with open(config["result"]) as f:
        result_json = json.load(f)

    raw_repeated = collect_repeated_indices(result_json)
    repeated = _consolidate_repeated(
        raw_repeated,
        max_unique_gap=config.get("max_unique_gap", 3),
        min_repeated_island=config.get("min_repeated_island", 2),
    )
    print(
        f"{len(raw_repeated)} raw / {len(repeated)} consolidated / {total} total entries",
        file=sys.stderr,
    )

    segments = []
    if total > 0:
        seg_start = 0
        seg_repeated = 0 in repeated

        for i in range(1, total):
            is_rep = i in repeated
            if is_rep != seg_repeated:
                segments.append(_build_segment(lookup, seg_start, i - 1, seg_repeated))
                seg_start = i
                seg_repeated = is_rep
        segments.append(_build_segment(lookup, seg_start, total - 1, seg_repeated))

    print(json.dumps(segments, indent=2, ensure_ascii=False))


def run_enrich(config: dict[str, Any]) -> None:
    lookup = build_entry_lookup(config["source"])
    print(f"Loaded {len(lookup)} entries from source for enrichment lookup", file=sys.stderr)

    with open(config["result"]) as f:
        result_json = json.load(f)

    # Inject total_duration_secs at top level
    if isinstance(result_json, dict) and lookup:
        first = lookup[0]
        last = lookup[-1]
        if first.get("start_ms") is not None and last.get("end_ms") is not None:
            result_json["total_duration_secs"] = (last["end_ms"] - first["start_ms"]) / 1000.0

    enrich_value(result_json, lookup)
    print(json.dumps(result_json, indent=2, ensure_ascii=False))
